tensor_comprehensions/tc_unit.py

```diff
@@ -0,0 +1,674 @@
+        number_elites=1,
+        threads=8,
+        gpus="0",
+        proto="/tmp/tuner.txt",
+        restore_from_proto=False,
+        restore_number=10,
+        log_generations=False,
+        tuner_min_launch_total_threads=64,
+        **kwargs
+    ):
+        # tuner_cache will look like:
+        # hash_key -> {"forward": options1, "backward": options2}
+        self.tuner_cache = {}
+        self.kwargs = kwargs
+        self.tc_lang = tc_lang
+        self.autotuner = ATenAutotuner(tc_lang)
+        self.set_autotuner_settings(
+            pop_size, crossover_rate, mutation_rate, generations, number_elites,
+            threads, gpus, proto, restore_from_proto, restore_number,
+            log_generations, tuner_min_launch_total_threads
+        )
+
+    def set_autotuner_settings(
+        self, pop_size, crossover_rate, mutation_rate, generations, number_elites,
+        threads, gpus, proto, restore_from_proto, restore_number, log_generations,
+        tuner_min_launch_total_threads,
+    ):
+        self.autotuner.pop_size(pop_size)
+        self.autotuner.crossover_rate(crossover_rate)
+        self.autotuner.mutation_rate(mutation_rate)
+        self.autotuner.generations(generations)
+        self.autotuner.number_elites(number_elites)
+        self.autotuner.threads(threads)
+        self.autotuner.gpus(gpus)
+        self.autotuner.proto(proto)
+        self.autotuner.restore_from_proto(restore_from_proto)
+        self.autotuner.restore_number(restore_number)
+        self.autotuner.log_generations(log_generations)
+        self.autotuner.tuner_min_launch_total_threads(tuner_min_launch_total_threads)
+
+    # We need to pass the inputs so that we can load the correct options from
+    # the cache that correspond to the inputs sizes. This is useful when the
+    # cache may contain multiple kernels and multiple sizes for each kernel
+    def load(self, filename, tc_name, inputs, num_candidates=1):
+        best_options = self.autotuner.load(filename, tc_name, inputs, num_candidates)
+        if num_candidates == 1:
+            return best_options[0]
+        return best_options
+
+    # if the cache_file is not "" then the tuning results would be saved to file
+    def tune_and_store(self, tc_name, inputs, mapping_options, cache_file=""):
+        options = mapping_options
+        if not isinstance(options, Options):
+            options = Options(options)
+        try:
+            best_options = self.autotuner.tune(cache_file, tc_name, inputs, options, [options])
+            return best_options
+        except RuntimeError:
+            return options
+
+    def autotune(self, *inputs, **kwargs):
+        input_tensors = get_tensors(list(inputs))
+        kwargs.update(self.kwargs)
+        name, backward_name = get_tc_names_from_kwargs(**kwargs)
+        kwargs.pop("name", None)
+        backward = True if backward_name is not None else False
+        hash_key = get_tc_hash_key(name, *input_tensors)
```

Comment:
So with this you are actually signing up for trouble as there is no guarantee that what you cache is correct in the long run. 

This should work fine for what you are currently doing with it but you should plan to pybind export functions (that we will need to expose) from [here](https://github.com/facebookresearch/TensorComprehensions/blob/master/include/tc/core/compilation_cache.h).

The reason is that the caching logic will depend on output tensor sizes, input/output tensor striding, allocation alignment many other fun things as we start digging deeper into optimizations.

There is nothing to change for now but it is just a friendly warning that you should not rely too much on your own framework-specific caching mechanism since this seems quite central to your autotuner wrapper logic.

Reply:
there is no support for strided tensors at the moment. When that is added and the task is closed for that, the python API will be modified accordingly.